lib/logstash/outputs/elasticsearch/template_manager.rb

```diff
@@ -46,15 +46,30 @@ def self.add_ilm_settings_to_template(plugin, template)
       # definition - remove any existing definition of 'template'
       template.delete('template') if template.include?('template') if plugin.maximum_seen_major_version < 8
       template['index_patterns'] = "#{plugin.ilm_rollover_alias}-*"
-      settings = template_settings(plugin, template)
+      settings = resolve_template_settings(plugin, template)
       if settings && (settings['index.lifecycle.name'] || settings['index.lifecycle.rollover_alias'])
         plugin.logger.info("Overwriting index lifecycle name and rollover alias as ILM is enabled")
       end
       settings.update({ 'index.lifecycle.name' => plugin.ilm_policy, 'index.lifecycle.rollover_alias' => plugin.ilm_rollover_alias})
     end
 
-    def self.template_settings(plugin, template)
-      plugin.maximum_seen_major_version < 8 ? template['settings']: template['template']['settings']
+    def self.resolve_template_settings(plugin, template)
+      return composable_index_template_settings(template) if template.key?('template')
+      return legacy_index_template_settings(template) if template.key?('settings')
+      return template_endpoint(plugin) == INDEX_TEMPLATE_ENDPOINT ?
+               composable_index_template_settings(template) :
+               legacy_index_template_settings(template)
```

Comment:
There is a lot of logic and guesswork here and I'm a bit concerned with the lack of logging
```suggestion
      if template.key?('template')
        logger.trace("template_settings: settings location for template is under 'template' key", :template => template, :template_api => plugin.template_api, :es_version => plugin.maximum_seen_major_version)
        composable_index_template_settings(template)
      elsif template.key?('settings')
        logger.trace("template_settings: settings location for template is at object's root", :template => template, :template_api => plugin.template_api, :es_version => plugin.maximum_seen_major_version)
        legacy_index_template_settings(template) 
      else
        template_endpoint = template_endpoint(plugin)
        logger.trace("template_settings: template doesn't have 'settings' or 'template' fields, falling back to auto detection, :template => template, :template_api => plugin.template_api, :es_version => plugin.maximum_seen_major_version, :detected_endpoint => template_endpoint)
        template_endpoint == INDEX_TEMPLATE_ENDPOINT ?
               composable_index_template_settings(template) :
               legacy_index_template_settings(template)
     end
```

Reply:
Added the logs between as well.